feat: add dynamic localisation and EDMC version compatibility

GLWine opened the pull request:
**Description:**

This pull request introduces a comprehensive refactor of the localisation utilities in `utils.py` to ensure robust compatibility with both legacy and future versions of EDMC, while improving code clarity, maintainability, and type safety.

---

**Key Changes:**

* **Unified Translation Interface:**
  Added `_get_tl_func`, which abstracts the selection of the correct translation function and translations object based on the detected EDMC version. This function returns both the appropriate translation callable (using `functools.partial` to pass context) and the correct translations object, allowing the rest of the codebase to interact with localisation in a consistent way regardless of the underlying EDMC API changes, especially around version 5.11.0 where the translation API shifted.

* **Future-Proof Localisation Logic:**
  All direct references to `l10n.Translations` and related constants have been replaced with the dynamic `translations_obj` returned by `_get_tl_func`. This ensures that the code will continue to work seamlessly when the legacy API is removed or made private in future EDMC versions (e.g., 6.0+). The translation wrapper function `__` now handles edge cases like empty, `None`, or fallback languages gracefully, returning the original string as a safe fallback and logging appropriate warnings and errors.

* **Refactored Localisation Functions:**
  The `__` and `available_langs` functions have been updated to use `translations_obj` and the new version detection logic, ensuring correct behaviour and compatibility across all supported EDMC versions. `available_langs` also conditionally uses either `str` or `Path` for the localisation directory path depending on the EDMC version for maximum compatibility.

* **Improved Type Annotations:**
  Type annotations throughout the module have been enhanced, replacing generic use of `any` with `Any` from the `typing` module and adding explicit annotations for functions and variables. This helps improve static analysis and reduce subtle bugs.

* **Enhanced Documentation and Readability:**
  Docstrings have been revised and expanded for clarity and consistency, improving maintainability and easing understanding for current and future contributors. Logging messages have been made more informative and consistent across the module.

* **General Code Cleanup:**
  Additional imports for `Callable`, `Any`, and `Tuple` from `typing` have been added where needed. Helper functions such as `get_by_path`, `human_format`, `is_number`, and `add_dicts` remain unchanged in functionality but benefit from improved type hints and documentation. Recursive dictionary merging in `add_dicts` ensures numeric values are summed correctly without modifying input dictionaries.

---

**Rationale:**

These changes address the evolving localisation API in EDMC, which is transitioning from the legacy `l10n.Translations` interface to a new, more flexible system. By centralising version detection and dynamically selecting the appropriate translation function and object, the codebase becomes more resilient to upstream changes and easier to maintain. The improved type annotations and comprehensive documentation further enhance code quality and developer experience.

This refactor guarantees that the plugin remains fully functional across all supported versions of EDMC, including future releases, and simplifies the eventual removal of legacy compatibility code when older versions are no longer supported.

---

**How to Test:**

* Run the plugin with multiple versions of EDMC, including versions before and after 5.11.0, to verify localisation works correctly in all cases.
* Verify language selection, translation overrides, and language listing behave as expected.
* Check logs to ensure warnings and errors related to missing translations or invalid language codes are properly emitted.
* Review the code for clarity, correct typing, and adherence to the new abstraction.

---

**Summary:**

This pull request modernises the localisation utilities by introducing a robust abstraction layer over EDMC’s translation API. It improves maintainability, compatibility, and prepares the plugin for upcoming EDMC changes, while preserving backward compatibility. The codebase has been expanded to accommodate this abstraction, ensuring that it remains stable against future up

---

**Note:**
The intent was not to revolutionize the original code, but to make it as robust as possible against future changes in EDMC’s localisation API. Even if this PR is not accepted as-is, I hope the underlying concept—abstracting and future-proofing the localisation logic—can be useful for the project going forward.
The codebase has grown in volume and complexity solely to introduce this abstraction and to ensure that it is not easily breakable by upstream changes.

GLWine committed:
feat: add EDMC version compatibility for l10n

Refactor localisation utilities for EDMC version compatibility and type safety

- Added `_get_core_version` to robustly determine the EDMC core version, supporting both string and callable appversion formats.
- Introduced `_get_tl_func` to abstract the selection of the correct translation function and translations object, ensuring compatibility with both legacy (`l10n.Translations`) and new (`l10n.translations`) APIs.
- Replaced all direct usage of `l10n.Translations` and related constants with the dynamic `translations_obj` returned by `_get_tl_func`, making the code future-proof for EDMC 6.0+.
- Updated the `__` and `available_langs` functions to use `translations_obj` and the new version logic.
- Improved type annotations throughout the module, replacing `any` with `Any` from the `typing` module.
- Added missing imports for `Callable`, `Any`, and `Tuple` from `typing`.
- Enhanced docstrings for clarity and PEP 257 compliance.
- General code cleanup and refactoring for maintainability and clarity.

aussig commented:
This is great, thank you.

I would like to suggest a further improvement though - since EDMC v5.11, the `translate` function has taken a further optional `lang` argument as its last parameter. As seen in the comments in `utils.py` on lines 91 and 92, I was intending to eliminate our custom `__(string: str, lang: str)` function altogether, replacing it with a simple call to `translate` but passing in the language.

Would you be able to modify your PR to incorporate this, I suggest returning three values from your `_get_tl_func()` instead of two: `_, __, translations_obj`.

GLWine commented:
But I have a question, if I have to manage what happens with lines https://github.com/GLWine/BGS-Tally-Fork/blob/86019ad0351dff915517f8742e35d1b73d078e86/bgstally/utils.py#L91-L92
now I wonder all the management from line https://github.com/GLWine/BGS-Tally-Fork/blob/86019ad0351dff915517f8742e35d1b73d078e86/bgstally/utils.py#L95-L123 should be removed?

could you ask me the thing?

aussig commented:
> now I wonder all the management from line https://github.com/GLWine/BGS-Tally-Fork/blob/86019ad0351dff915517f8742e35d1b73d078e86/bgstally/utils.py#L95-L123 should be removed?

Yes, that was the plan, but of course it will need testing to make sure the plan actually works :)

GLWine commented:
I thought to solve the problem like this:
```py
def __(string: str, lang: str) -> str:
    """
    Translate a string using the specified language, compatible with all EDMC versions.

    Args:
        string (str): The string to translate.
        lang (str): The language code to use for translation.

    Returns:
        str: The translated string.
    """
    return translations_obj.translate(string, context=__file__, lang=lang)
```
I can't add it to the tuple because the language must be managed dynamically, every time functool.partial is activated it assigns the parameters at that moment and I can't do it later.
So I have to manage it anyway as a special function, extremely simplified

aussig commented:
That documentation is just meant to be an example.  I know this because I wrote it :D

My addition of the `lang` parameter to the EDMC `translate` function a year ago was specifically to support this 'alternate language' feature in BGS-Tally, but of course it takes time for new additions like that to percolate through the EDMC release cycle, hence the addition of `__()` as a function in our code which replicated the EDMC code (and which was always meant to be temporary until we could rely on the feature being present in core EDMC).

I'm pretty sure (but not certain) that I tested the `__ = functools.partial(l10n.Translations.translate, context=__file__, lang=lang)` approach back then, but am very willing to accept that I didn't and that it therefore doesn't work!  Have you tried extending the partial to include a `lang` parameter?

GLWine commented:
we can use this structure
__ = functools.partial(l10n.Translations.translate, context=__file__, lang=lang)

but `lang` must be defined or taken from somewhere, this means that it must update with the config update for example

GLWine commented:
I should study better how languages ​​are managed in the config for example, provided that you only use it in webooks... otherwise I would leave the very simple def __(...)

GLWine commented:
To keep this in mind, I also submitted a possible solution to the problem encountered in EDMC 
- https://github.com/EDCD/EDMarketConnector/issues/2437#issuecomment-2914564558

aussig commented:
Yes I'm happy with that - as you say we need to catch that exception for the moment, until the issue is resolved in EDMC core, so a `__()` wrapper function on our side makes sense.

GLWine commented:
@aussig what do you think about just putting the English language as a file? 
Simply rename the template end to `en.strings`, then we might not have to handle the exception!

What do you think?

aussig commented:
Yes that could work (would need testing of course).

There is already `en.template.strings` which could "simply" be renamed, but note that this file is used by Crowdin as the master source file for English strings, so we would have to synchronise a change to `crowdin.yaml` (line 2) to change this in the same commit.  I hope Crowdin wouldn't get confused by that...

GLWine commented:
I tested the direct approach using a tuple, but it turned out to be significantly more complex to manage: the `KeyError` is raised again, this time by the `available_langs()` function. The issue stems from the fact that this function retrieves language codes from the plugin’s `L10n` folder and then attempts to match them with the available contents in EDMC via `l10n.Translation.contents()`. However, since the `en.strings` file is not present in EDMC, the error is triggered.

Personally, I believe that—for now—the most practical solution is to implement a wrapper function that explicitly handles the special case of the `en` language. Once EDMC provides a structural fix for this issue, we can consider replacing the wrapper with a more integrated tuple-based approach.

GLWine committed:
refactor: improve translation wrapper and language listing

- Refactored the `__` translation wrapper to improve error handling and logging:
  - Now logs a warning if translation is requested with empty or None language.
  - Returns the original string for empty, None, or fallback language without attempting translation.
  - Catches KeyError during translation and logs a detailed error, returning the original string as fallback.
- Improved docstrings for all utility functions to be more descriptive and PEP 8/PEP 257 compliant.

GLWine commented:
### Detailed description (https://github.com/aussig/BGS-Tally/pull/324/commits/cd30bd24b9a9298cece199762658c6a3c4616447)

- **Refactor of the `__` translation function:**
    - The function now handles more robustly the cases where the requested language is empty (`""`), `None` or matches the fallback language (usually `'en'`), returning the original string directly without attempting translation.
    - A warning is added to the log if translation is requested with empty language or `None`, making it easier to debug incorrect calls.
    - If a `KeyError` is raised during translation (e.g. missing translation file or unavailable language), the error is logged in detail and the function returns the original string as a fallback, avoiding crashes or unhandled exceptions.

- **Docstring improvements:**
    - All utility functions have been updated with docstrings, improving the readability and maintainability of the code.

- **Overall impact:**
    - The code is now more robust, clear, and easily extendable.
    - Error handling in localization is safer and more informative, reducing the possibility of silent errors or unexpected crashes.
    - Inline documentation is more complete, making it easier for future developers or maintainers of the project to understand.

---

**In summary:**
This commit improves the quality and robustness of the translation and language management utilities, making the code more reliable and documented.

aussig commented on bgstally/utils.py:
```diff
@@ -15,39 +16,116 @@
 
-# Localisation conditional translation function for when PR [2188] is merged in to EDMC
-# __ = functools.partial(l10n.Translations.translate, context=__file__, lang=lang)
+def _get_core_version(appversion) -> semantic_version.Version:
```
I think `_get_edmc_version()` or `_get_core_edmc_version()` would be clearer here

GLWine replied:
Risolto in: https://github.com/aussig/BGS-Tally/pull/324/commits/8e1b3cda675a4233baef5197ece83333c0881029 utilizzando `_get_edmc_version()`

aussig commented on bgstally/utils.py:
```diff
@@ -15,39 +16,116 @@
+
+
+# Holds the parsed core version of the application.
+core_version: semantic_version.Version = _get_core_version(appversion)
```
Perhaps rename to `edmc_version` or `core_edmc_version` for clarity.

GLWine replied:
Risolto in: https://github.com/aussig/BGS-Tally/pull/324/commits/8e1b3cda675a4233baef5197ece83333c0881029 utilizzando `edmc_version`

GLWine committed:
refactor: update `core` to `edmc` for clarity

- Renamed _get_core_version() to _get_edmc_version() and core_version to edmc_version as requested by @aussig, for clarity and consistency with EDMC terminology.
- Updated all references and docstrings to reflect the new naming.
- No functional changes to version parsing logic; only naming and documentation improved.

GLWine commented:
Before closing the PR I would like to discuss the control of the pre 5.0.0 version.
I tried to start the BGS-Tally in version 4.2.7, the latter did not start due to other incompatibilities that are not managed.

At this point I could lighten the code by directly excluding its management, what do you think @aussig?

GLWine committed:
refactor: use appversion() directly

- Removed the _get_edmc_version() function and related logic.
- Now assigns edmc_version directly from appversion(), simplifying version handling.
- Updated all references and docstrings accordingly.
- No changes to translation, language utilities, or other helper functions.

GLWine commented:
At this point, the `edmc_version` variable is assigned by the `appversion()` function (https://github.com/aussig/BGS-Tally/pull/324/commits/03152614842aa050b24a29b5967b240b6c8058f1).
- Simplifying the structure and logic.

If for you @aussig there are no further points to validate, you could also call this PR concluded.
